### DISCUSSION TITLE
Need assistance with multiple ports, health checks and out pipeline

### DISCUSSION THREAD

**Discussion by kingDaddyFreshness** [open-telemetry/opentelemetry-collector#6883] [General]
Hello all! 

as you already know, the open-telemetry collector can be configured to use different ports for different protocols (one for http, one for grpc, etc).  Additionally, there is a different port for the health check.  For reasons that I won't bore you with, our deployment pipeline is not great at handling an application with multiple ports - particularly in regards to health checks.  When we provision the collector, our pipeline "assumes" that the service (ie. http or grpc) is running on the same port as the health check. 
 I'm wondering if there is a simple solution here?

Here are the solutions I'm imagining:
1. we fix our pipeline so that it doesn't assume the "service" isn't on the same port as the health check.  Unfortunately I have no control over this and I do not own the pipeline.
2. we wait for someone else to modify the collector so that the health check can be on the same port as the service.  We'd then use one endpoint (at port X) for the service and one endpoint (at port X) for the health check.  I'm not sure if this is even in the works or if others are already working on it?
3. I learn the codebase of the collector and create my own custom version that uses different endpoints on the same port for different things.  I've got no experience with Go and am not sure about licensing/legality of making custom collector.  I also don't like this option because we then diverge from the primary code base.  I supposed I could do the work as a pull request and cross my fingers?
4. Other ideas?

I've seen in other forums that other folks have run into this problem.  Would very much like some guidance on a solution.  Thanks - and let me know if you have any questions!

-john

**Comment by jpkrohling** [comment]
> open-telemetry collector can be configured to use different ports for different protocols (one for http, one for grpc, etc). Additionally, there is a different port for the health check

We did talk about this briefly before in different flavors, including when talking about the Jaeger remote sampling extension clashing with the Jaeger receiver (cc @frzifus).

One possible solution would be to add an option for users to share the same port and let components reuse it, registering their endpoints. As far as I know, nobody played with this before and nobody is planning on working on it.

**Reply by frzifus** [reply]
I moved some code around to see how this could work. But its harder then i thought. If we only would support grpc it shouldnt be to hard.
Would affect https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14707.
